src/Castle.Facilities.AspNetCore.Tests/WindsorRegistrationExtensionsTestCase.cs

```diff
@@ -0,0 +1,354 @@
+		[TestCase(typeof(CompositeController))]
+		[TestCase(typeof(CompositeViewComponent))]
+		public void Should_resolve_Composite_Singleton_CrossWired_from_ServiceProvider(Type compositeType)
+		{
+			testContext.WindsorContainer.Register(Component.For(compositeType).CrossWired().LifestyleSingleton());
+
+			Assert.DoesNotThrow(() =>
+			{
+				using (var sp = testContext.ServiceCollection.BuildServiceProvider())
+				{
+					sp.GetRequiredService(compositeType);
+				}
+			});
+		}
+
+		[TestCase(typeof(CompositeTagHelper))]
+		[TestCase(typeof(CompositeController))]
+		[TestCase(typeof(CompositeViewComponent))]
+		public void Should_resolve_Composite_Scoped_CrossWired_from_ServiceProvider(Type compositeType)
+		{
+			testContext.WindsorContainer.Register(Component.For(compositeType).CrossWired().LifestyleScoped());
+
+			Assert.DoesNotThrow(() =>
+			{
+				using (var sp = testContext.ServiceCollection.BuildServiceProvider())
+				{
+					sp.GetRequiredService(compositeType);
+				}
+			});
+		}
+
+		[TestCase(typeof(CompositeTagHelper))]
+		[TestCase(typeof(CompositeController))]
+		[TestCase(typeof(CompositeViewComponent))]
+		public void Should_resolve_Composite_Transient_CrossWired_from_ServiceProvider(Type compositeType)
+		{
+			testContext.WindsorContainer.Register(Component.For(compositeType).CrossWired().LifestyleTransient());
+
+			Assert.DoesNotThrow(() =>
+			{
+				using (var sp = testContext.ServiceCollection.BuildServiceProvider())
+				{
+					sp.GetRequiredService(compositeType);
+				}
+			});
+		}
+
+		[TestCase(LifestyleType.Bound)]
+		[TestCase(LifestyleType.Custom)]
+		[TestCase(LifestyleType.Pooled)]
+		[TestCase(LifestyleType.Thread)]
+		//[TestCase(LifestyleType.Undefined)] // Already throws System.ArgumentOutOfRangeException: Undefined is not a valid lifestyle type 
+		public void Should_throw_if_CrossWired_with_these_LifestyleTypes(LifestyleType unsupportedLifestyleType)
+		{
+			Assert.Throws<NotSupportedException>(() =>
+			{
+				var componentRegistration = Component.For<AnyComponent>().CrossWired();
+				componentRegistration.LifeStyle.Is(unsupportedLifestyleType);
+				if (unsupportedLifestyleType == LifestyleType.Custom)
+				{
+					componentRegistration.LifestyleCustom<AnyComponentWithLifestyleManager>();
+				}
+				testContext.WindsorContainer.Register(componentRegistration);
+			});
+		}
+
+		[Test]
+		public void Should_throw_if_Facility_is_added_without_calling_CrossWiresInto_on_IWindsorContainer_AddFacility()
+		{
+			using (var container = new WindsorContainer())
+			{
+				Assert.Throws<InvalidOperationException>(() =>
+				{
+					container.AddFacility<AspNetCoreFacility>();
+				});
+			}
+		}
+
+		[Test]
+		public void Should_resolve_IMiddleware_from_Windsor()
+		{
+			testContext.ApplicationBuilder.UseMiddlewareFromWindsor<AnyMiddleware>(testContext.WindsorContainer);
+
+			Assert.DoesNotThrow(() => { testContext.WindsorContainer.Resolve<AnyMiddleware>(); });
+		}
```

Comment:
Found more sucky API's. 

@jonorossi Do you have any tips here?

Reply:
Would be great of we could get rid of the IWindsorContainer parameter here. It is a bit late so I will have to pick this up tomorrow.

Reply:
Not knowing the ASP.NET Core stuff, I guess you'd have to store the container reference and get access via `IApplicationBuilder`.

Reply:
@jonorossi 

Correct. I managed to take advantage of the facility to do this. Here is some sample code for new usage:

```csharp
public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
{
	// For making component registrations of middleware easier
	Container.GetFacility<AspNetCoreFacility>().RegistersMiddlewareInto(app);

	// Add custom middleware, do this if your middleware uses DI from Windsor
	Container.Register(Component.For<CustomMiddleware>().DependsOn(Dependency.OnValue<ILoggerFactory>(loggerFactory)).LifestyleScoped().AsMiddleware());
	
	// ... 
}
```

This now means we have first class support for Middleware registration(including lifestyles) using component registrations as opposed to the `roll your own` approach which coerced everything to singleton. 

I will present this change as a separate commit once I heard back from @xhafan on issue #411 to make the review a little easier and squish once we are all happy with the proposed change.

Reply:
Man, sorting out the memory management was an absolute bitch. These were pre-existing problems where CrossWired components were consuming each other (https://github.com/fir3pho3nixx/Windsor/pull/25).